kotlinpoet/src/main/java/com/squareup/kotlinpoet/TypeName.kt

```diff
@@ -95,12 +95,23 @@ public sealed class TypeName constructor(
 
   override fun equals(other: Any?): Boolean {
     if (this === other) return true
-    if (other == null) return false
-    if (javaClass != other.javaClass) return false
-    return toString() == other.toString()
+    if (javaClass != other?.javaClass) return false
```

Comment:
Thanks, so I looked it up and it seems the check uses the runtime class. 

https://github.com/JetBrains/kotlin/blob/30788566012c571aa1d3590912468d1ebe59983d/libraries/stdlib/jvm/runtime/kotlin/jvm/JvmClassMapping.kt#L85

So when comparing two `ClassName`, the equals in `ClassName` will call `super.equals()` which will invoke `TypeName.equals()`. Because the runtime class of both is `ClassName`, the `super.equals()` will return true if nullable, annotations, and tags are the same.

Reply:
Ah, that makes sense, thank you for clarifying!